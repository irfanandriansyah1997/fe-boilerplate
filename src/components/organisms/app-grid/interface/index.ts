import { FC } from 'react';

import { ICellProps, IGridProps } from '../../../molecules/grid/interface';

/**
 * App Grid Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.30
 */
export interface IAppGridProps extends IGridProps {
  Cell: FC<ICellProps>;
  onForceRerender(): void;
  onUpdateGrid(): void;
}
