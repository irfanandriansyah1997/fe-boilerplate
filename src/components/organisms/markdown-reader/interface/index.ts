/**
 * Markdown Content Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IMarkdownContentProps {
  content: string;
}

/**
 * Markdown Hooks Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IMarkdownHooks {
  content?: string;
  status: number;
  url: string;
}

/**
 * Markdown Reader Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IMarkdownReaderProps {
  markdownUrl: string;
}
