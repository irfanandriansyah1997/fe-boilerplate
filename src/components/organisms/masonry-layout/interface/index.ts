import { FC } from 'react';

/**
 * Grouping Masonry Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IGroupingMasonry {
  item: IMasonryItem[];
  key: string;
}

/**
 * Masonry Hooks Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.24
 */
export interface IMasonryHooks {
  gridCount: number;
  item: IGroupingMasonry[];
}

/**
 * Masonry Item Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IMasonryItem {
  item: Record<string, any>;
  key: string;
}

/**
 * Masonry Layout Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.15
 */
export interface IMasonryLayout {
  component: FC<any>;
  item: IMasonryItem[];
  maxGrid: number;
}
