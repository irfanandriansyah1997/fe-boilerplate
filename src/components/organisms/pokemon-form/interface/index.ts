/**
 * Pokemon Form Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.19
 */
export interface IPokemonFormProps {
  initialPokemonName?: string;
  onSubmit(pokemonName: string): void;
  pokemonName: string;
}

/**
 * Pokemon Option Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.19
 */
export interface IPokemonOptionProps {
  onClick(pokemonName: string): void;
  pokemonName: string;
}
