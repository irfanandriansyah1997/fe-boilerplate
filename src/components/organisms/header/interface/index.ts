import { IMenuItem } from '../../../../interface/component';

/**
 * Header Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.14
 */
export interface IHeaderProps {
  menu: IMenuItem[];
}
