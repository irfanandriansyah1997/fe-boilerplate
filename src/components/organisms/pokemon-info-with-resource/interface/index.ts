import { IPokemonCachePayload } from '../../../../section/suspense-api/suspense-image/interface';
import { IPokemonCardTypeEnum } from '../../../molecules/pokemon-card/interface';

/**
 * Pokemon Info With Resource Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.11.19
 */
export interface IPokemonInfoWithResourceProps {
  resource: IPokemonCachePayload;
  type?: IPokemonCardTypeEnum;
}
