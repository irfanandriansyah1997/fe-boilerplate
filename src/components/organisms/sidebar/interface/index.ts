import { IMenuItem } from '../../../../interface/component';

/**
 * Sidebar Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.14
 */
export interface ISidebarProps {
  menu: IMenuItem[];
  subtitle: string;
  title: string;
}
