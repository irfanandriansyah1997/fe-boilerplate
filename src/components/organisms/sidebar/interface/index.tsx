import { IMenu } from '../../../../interface/component';

/**
 * Sidebar Mobile Menu Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.24
 */
export interface ISidebarMobilMenuProps extends ISidebarProps {
  show: boolean;
  toggleShow(show: boolean): void;
}

/**
 * Sidebar Props Interface
 * @author Irfan Andriansyah <irfan@99.co>
 * @since 2021.10.22
 */
export interface ISidebarProps extends IMenu {
  onClickBack(): void;
}
