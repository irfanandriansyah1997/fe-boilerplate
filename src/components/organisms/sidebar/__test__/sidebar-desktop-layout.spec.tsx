import { fireEvent, render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import { IMenuItem } from '../../../../interface/component';
import SidebarDesktopLayout from '../section/desktop-layout.component';

const MOCK_MENU: IMenuItem[] = [
  {
    text: `Wrap Use Memo`,
    to: `/context/part-1`
  },
  {
    text: `Separate Context Value`,
    to: `/context/part-2`
  },
  {
    text: `Colocate State`,
    to: `/context/part-3`
  }
];

describe(`Testing Sidebar On Desktop Site`, () => {
  it(`Snapshot Testing`, () => {
    const onClickBackSpy = jest.fn();
    const onClickItemSpy = jest.fn();
    const { getByTestId } = render(
      <MemoryRouter initialEntries={[`/context/part-1`]}>
        <SidebarDesktopLayout
          onClickBack={onClickBackSpy}
          onClickItem={onClickItemSpy}
          subtitle="Sample Subtitle Course"
          title="Title Course"
          menu={MOCK_MENU}
        />
      </MemoryRouter>
    );

    expect(getByTestId(/^sidebar-dsite$/i)).toBeInTheDocument();
    expect(getByTestId(/^sidebar-dsite$/i)).toMatchSnapshot();
  });

  it(`Testing Element HTML On Sidebar Desktop Layout`, async () => {
    const onClickBackSpy = jest.fn();
    const onClickItemSpy = jest.fn();
    const { findAllByTestId, getByText } = render(
      <MemoryRouter initialEntries={[`/context/part-1`]}>
        <SidebarDesktopLayout
          onClickBack={onClickBackSpy}
          onClickItem={onClickItemSpy}
          subtitle="Sample Subtitle Course"
          title="Title Course"
          menu={MOCK_MENU}
        />
      </MemoryRouter>
    );

    expect(getByText(/Title Course/)).toBeInTheDocument();
    expect(getByText(/Sample Subtitle Course/)).toBeInTheDocument();

    const element = await findAllByTestId(/sidebar-dsite-item/);
    expect(element.length).toBe(3);

    expect(element[0].textContent).toBe(`doneWrap Use Memo`);
    expect(element[1].textContent).toBe(`doneSeparate Context Value`);
    expect(element[2].textContent).toBe(`doneColocate State`);
  });

  it(`Simulate User Click Back Button`, async () => {
    const onClickBackSpy = jest.fn();
    const onClickItemSpy = jest.fn();
    const { getByTestId } = render(
      <MemoryRouter initialEntries={[`/context/part-1`]}>
        <SidebarDesktopLayout
          onClickBack={onClickBackSpy}
          onClickItem={onClickItemSpy}
          subtitle="Sample Subtitle Course"
          title="Title Course"
          menu={MOCK_MENU}
        />
      </MemoryRouter>
    );

    /**
     * Simulate User Click Back Button
     */
    expect(getByTestId(/sidebar-dsite-click-back/)).toBeInTheDocument();
    fireEvent.click(getByTestId(/sidebar-dsite-click-back/));
    expect(onClickBackSpy).toHaveBeenCalled();
  });

  it(`Simulate User Click Sidebar Item`, async () => {
    const onClickBackSpy = jest.fn();
    const onClickItemSpy = jest.fn();
    const { findAllByTestId } = render(
      <MemoryRouter initialEntries={[`/context/part-1`]}>
        <SidebarDesktopLayout
          onClickBack={onClickBackSpy}
          onClickItem={onClickItemSpy}
          subtitle="Sample Subtitle Course"
          title="Title Course"
          menu={MOCK_MENU}
        />
      </MemoryRouter>
    );

    const element = await findAllByTestId(/sidebar-dsite-item/);
    expect(element.length).toBe(3);

    /**
     * Simulate User Click Sidebar Item
     */
    fireEvent.click(element[1]);
    expect(onClickItemSpy).toHaveBeenCalledWith({
      isPrimary: undefined,
      text: `Separate Context Value`,
      to: `/context/part-2`
    });
  });

  it(`Simulate User Click Sidebar Item With OnClickItem Props Is Undefined`, async () => {
    const onClickBackSpy = jest.fn();
    const onClickItemSpy = jest.fn();
    const { findAllByTestId } = render(
      <MemoryRouter initialEntries={[`/context/part-1`]}>
        <SidebarDesktopLayout
          onClickBack={onClickBackSpy}
          subtitle="Sample Subtitle Course"
          title="Title Course"
          menu={MOCK_MENU}
        />
      </MemoryRouter>
    );

    const element = await findAllByTestId(/sidebar-dsite-item/);
    expect(element.length).toBe(3);

    /**
     * Simulate User Click Sidebar Item
     */
    fireEvent.click(element[1]);
    expect(onClickItemSpy).not.toHaveBeenCalled();
  });
});
