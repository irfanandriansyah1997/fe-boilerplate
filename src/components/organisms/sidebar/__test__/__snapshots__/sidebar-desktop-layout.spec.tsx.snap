// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Sidebar On Desktop Site Snapshot Testing 1`] = `
<div
  class="o-sidebar"
  data-testid="sidebar-dsite"
>
  <h6>
    <button
      class="material-icons"
      data-testid="sidebar-dsite-click-back"
      type="button"
    >
      arrow_back
    </button>
    Title Course
  </h6>
  <p>
    Sample Subtitle Course
  </p>
  <div
    class="o-sidebar__menu"
  >
    <a
      aria-current="page"
      class="o-sidebar__item o-sidebar__item--active"
      data-testid="sidebar-dsite-item"
      href="/context/part-1"
    >
      <span
        class="material-icons"
      >
        done
      </span>
      Wrap Use Memo
    </a>
    <a
      class="o-sidebar__item"
      data-testid="sidebar-dsite-item"
      href="/context/part-2"
    >
      <span
        class="material-icons"
      >
        done
      </span>
      Separate Context Value
    </a>
    <a
      class="o-sidebar__item"
      data-testid="sidebar-dsite-item"
      href="/context/part-3"
    >
      <span
        class="material-icons"
      >
        done
      </span>
      Colocate State
    </a>
  </div>
</div>
`;
